File: src_C/ice_setup.c
/* ice_setup.c -- Phase 0: build a fully-populated IceState. Ported from IceSetup.cs. Derivations use `double`,
 * which is bit-exact for setup. */
#include "ice.h"
#include <stdlib.h>
#include <math.h>

/* ICE quality: 0 = Normal (default), 1 = Fine. Set by openice.c's -fine. (See docs/parameters.md.) */
int ice_fine = 0;

/* Reconstruction target ("kind"): which scanner reconstruction profile to use. 8 = LS-5000 (default,
 * byte-exact); 7 and 9 are the two other targets. Set by openice.c's -kind. Only the ReconCoefBase matrix and
 * BandLookaheadRows differ between kinds; BandLookaheadRows is output-inert at the LS-5000's 4000/285 dpi
 * (dpi-vs-threshold resolves the same for all). */
int ice_kind = 8;

/* [kind index 0=7,1=8,2=9][channel R,G,B][IR-slope k + 3 bands x {loCoef,hiCoef}] -- exact 32-bit float patterns */
static const unsigned int recon_coef_by_kind[3][3][7] = {
  { /* kind 7 */
    {0x3F8CCCCDu,0x3FAE147Bu,0x3FA8F5C3u,0x3FAF5C29u,0x3FA66666u,0x3FAB851Fu,0x3FA00000u},
    {0x3F8CCCCDu,0x3FAF5C29u,0x3FA66666u,0x3FACCCCDu,0x3FA51EB8u,0x3FA66666u,0x3F9EB852u},
    {0x3F8CCCCDu,0x3FAB851Fu,0x3FA00000u,0x3FA8F5C3u,0x3FA00000u,0x3FA00000u,0x3F9AE148u} },
  { /* kind 8 (LS-5000, default) */
    {0x3F8CCCCDu,0x3F9AE148u,0x3F8B851Fu,0x3F95C28Fu,0x3F8A3D71u,0x3F851EB8u,0x3F75C28Fu},
    {0x3F8CCCCDu,0x3F9D70A4u,0x3F90A3D7u,0x3F91EB85u,0x3F866666u,0x3F6E147Bu,0x3F570A3Du},
    {0x3F8CCCCDu,0x3F90A3D7u,0x3F851EB8u,0x3F8A3D71u,0x3F828F5Cu,0x3F7851ECu,0x3F63D70Au} },
  { /* kind 9 */
    {0x3F800000u,0x400D70A4u,0x4005C28Fu,0x400AE148u,0x40051EB8u,0x40028F5Cu,0x3FFAE148u},
    {0x3F800000u,0x400EB852u,0x400851ECu,0x4008F5C3u,0x40033333u,0x3FF70A3Du,0x3FEB851Fu},
    {0x3F800000u,0x400851ECu,0x40028F5Cu,0x40051EB8u,0x400147AEu,0x3FFC28F6u,0x3FF1EB85u} }
};
static const int band_lookahead_by_kind[3] = { 950, 1600, 2500 };   /* kinds 7, 8, 9 */

/* map ice_kind {7,8,9} -> table index {0,1,2}; anything else -> 1 (kind 8 default) */
static int kind_index(void) { int i = ice_kind - 7; return (i < 0 || i > 2) ? 1 : i; }

static int    idx_of(float anchor, int maxIdx) { int i = (int)((double)anchor * (double)maxIdx); return i > maxIdx ? maxIdx : i; }
static float  lut_sub(const IceState *s, float anchor, int maxIdx) { return (float)((double)s->densityLut[idx_of(anchor, maxIdx)] - (double)s->densityLut[maxIdx]); }
static float  lut_at (const IceState *s, float anchor, int maxIdx) { return s->densityLut[idx_of(anchor, maxIdx)]; }
static float  bias_param(const IceState *s, float fv1, int maxIdx, double a0, double a4) {
    const double a9c = 2.0;
    if (fv1 <= a0) return (a4 < fv1) ? lut_sub(s, fv1, maxIdx) : 0.0f;
    if (a9c <= fv1) return (a4 < fv1) ? lut_sub(s, fv1, maxIdx) : 0.0f;
    { double g = a9c - (double)fv1;
      return (float)((double)s->densityLut[maxIdx] - (double)s->densityLut[idx_of((float)g, maxIdx)]); }
}

IceState *ice_setup_create(int W, int H, int dpi) {
    IceState *s = (IceState *)calloc(1, sizeof(IceState));
    (void)H;

    /* geometry / config */
    cfg_wi(s, Cfg_Dpi, dpi);
    cfg_wi(s, 0x010, 2);
    cfg_wi(s, Cfg_TriggerEdgeMode, 0);
    cfg_wi(s, Cfg_ReconIrRefAdjust, 1); cfg_wi(s, Cfg_ReconBand0, 1); cfg_wi(s, Cfg_ReconBand1, 1); cfg_wi(s, Cfg_ReconBand2, 1);
    /* Cfg_TargetId (0x3B4) is a MISNOMER: it is NOT the reconstruction kind. It reads 8 for ALL three kinds
     * 7/8/9 -- it is the give-up sample cap read as `maxBelow` in ice_back_trigger. Writing ice_kind here made
     * -kind 7 give up at >7 and -kind 9 at >9, flipping the give-up decision on a large fraction of pixels.
     * It is kind-invariant: always 8. */
    cfg_wi(s, Cfg_ChannelCount, 2); cfg_wi(s, Cfg_BandCount, 0); cfg_wi(s, Cfg_TargetId, 8);
    cfg_wi(s, 0xEC8, 4);
    cfg_wi(s, 0xECC, W + 8);
    cfg_wi(s, 0xED0, W + 4);
    cfg_wi(s, Cfg_ImageWidth, W);
    cfg_wi(s, Cfg_RowCounter, 0);
    cfg_wi(s, Cfg_ClampToL3Enable, 1); cfg_wi(s, Cfg_DrainFlag, 0); cfg_wi(s, Cfg_ClampToL3Flag, 0);

    cfg_wi(s, Cfg_WarmupRows, 7); cfg_wi(s, 0xE1C, 6); cfg_wi(s, 0xE28, 2); cfg_wi(s, 0xE3C, 3); cfg_wi(s, 0xE40, 1);
    cfg_wi(s, 0xE2C, 0); cfg_wi(s, 0xE30, 1); cfg_wi(s, 0xE34, 2);
    cfg_wu16(s, Cfg_MaxDensityIdx, 65535); cfg_wu16(s, 0xD72, 0);
    cfg_wi(s, Cfg_BandLookaheadRows, band_lookahead_by_kind[kind_index()]); cfg_wi(s, Cfg_WeightMinRows, 550); cfg_wi(s, 0x458, 1);

    /* density LUT: LUT[i] = 65535*log2(i+1)/16 */
    s->densityLut = (float *)malloc(65536 * sizeof(float));
    { double k = 65535.0 / (16.0 * log(2.0)); int i;
      for (i = 0; i <= 65535; i++) s->densityLut[i] = (float)(k * log((double)i + 1.0)); }

    /* true constants */
    cfg_wbits(s, 0x3BC, 0x00000000u);
    cfg_wbits(s, Cfg_WeightFloor, 0x3CA3D70Au);   /* 0.02 */
    cfg_wbits(s, Cfg_IRref, 0x00000000u);
    cfg_wbits(s, Cfg_MaskOverrideR, 0); cfg_wbits(s, Cfg_MaskOverrideG, 0); cfg_wbits(s, Cfg_MaskOverrideB, 0);
    /* Trigger probe offsets are 4 for every kind at 0xF70..0xF76. (The core carries a second copy at
     * 0x4A0/0x4A4 that reads 3 for kind 7, but mirroring it here measurably HURTS kind 7 -- not this. ) */
    cfg_wu16(s, Cfg_TriggerOffsetL, 4); cfg_wu16(s, Cfg_TriggerOffsetR, 4); cfg_wu16(s, 0xF74, 4); cfg_wu16(s, 0xF76, 4);

    /* per-channel dither amounts (0.015/0.015/0.025) -- LS-5000 target-8, scan-invariant. */
    cfg_wbits(s, Cfg_DitherAmtR, 0x3C75C28Fu);   /* 0.015 */
    cfg_wbits(s, Cfg_DitherAmtG, 0x3C75C28Fu);   /* 0.015 */
    cfg_wbits(s, Cfg_DitherAmtB, 0x3CCCCCCDu);   /* 0.025 */

    cfg_wbits(s, Cfg_StageGain2, 0x3FA00000u);   /* 1.25 */
    cfg_wbits(s, Cfg_StageGain1, 0x3FA00000u);   /* 1.25 */
    cfg_wbits(s, Cfg_StageGain0, 0x3FA00000u);   /* 1.25 (Normal uses 1.25, not 1.0) */
    cfg_wi(s, Cfg_ClampToL3Enable, 1);

    /* ICE Fine (`-fine`). Between Normal and Fine, the only reconstruction fields that change are the three
     * stage gains (1.25 -> 1.0) and the L3 clamp (on -> off). Setting just these reproduces Fine output
     * byte-exactly (70,542,642/70,542,642, 0 diff). (Fine also zeroes an unread 0xF50 word and a 192-float
     * table at 0x94; neither affects the reconstruction output.) */
    if (ice_fine || getenv("ICE_FINE")) {
        cfg_wbits(s, Cfg_StageGain2, 0x3F800000u); cfg_wbits(s, Cfg_StageGain1, 0x3F800000u); cfg_wbits(s, Cfg_StageGain0, 0x3F800000u);
        cfg_wi(s, Cfg_ClampToL3Enable, 0);
    }

    /* per-channel 7-float reconstruction coefficient block (IR-slope k + 3 bands x {loCoef,hiCoef}), selected
     * by the reconstruction target (ice_kind, default 8 = LS-5000), at config 0xED8..0xF28. */
    { const unsigned int (*pc)[7] = recon_coef_by_kind[kind_index()]; int c, i;
      for (c = 0; c < 3; c++) for (i = 0; i < 7; i++) cfg_wbits(s, Cfg_ReconCoefBase + c * 0x1C + i * 4, pc[c][i]); }

    /* per-scan parameters */
    { int chans = 2, maxIdx = 65535; double statA = 1.0, statB = 0.0;
      float f30 = lut_sub(s, 0.85f, maxIdx);
      /* The weight-ramp anchor is per-kind: config 0xF30 is c4701ac0 (-960.418) for kinds 7/8 but c4702180
       * (-960.523) for kind 9. WeightSlope = 1/f30 follows (k8 ba887952, k9 ba88757c), so overriding f30 here
       * reproduces the whole ramp. */
      if (ice_kind == 9) f30 = float_from_bits(0xc4702180u);
      float f34 = lut_sub(s, 1.0f, maxIdx);
      float f40 = bias_param(s, 0.98f, maxIdx, statA, statB);
      float f4c = lut_at(s, 0.065f, maxIdx);
      cfg_wbits(s, 0xF30, bits_of_float(f30));
      cfg_wbits(s, 0xF34, bits_of_float(f34));
      { double fv5 = (double)f30 - (double)f34;
        if (fv5 == statB) { cfg_wf(s, Cfg_WeightSlope, -(float)maxIdx); cfg_wf(s, Cfg_WeightRamp, (float)statA); }
        else { double slope = statA / fv5; cfg_wf(s, Cfg_WeightSlope, (float)slope); cfg_wf(s, Cfg_WeightRamp, (float)(slope * (double)f30)); } }
      cfg_wf(s, Cfg_WeightBias, f40);
      cfg_wf(s, Cfg_MaxDensitySum, (float)((float)chans * (float)maxIdx));
      cfg_wf(s, 0xF48, 0.0f);
      cfg_wf(s, Cfg_DustFloor, f4c); }

    /* fixed core constants */
    s->Zero               = 0.0f;
    s->RConfGain          = 2.0f;
    s->MaskClampHi        = 1.0f;
    s->DitherFloor        = 0.0f;
    s->DitherEnvScale     = 4.0f;
    /* FIX 5 -- the dither BAND is per-kind. The original carries separate dither constant blocks; the two band
     * anchors differ by kind:
     *     kind 7:      0.04 / 0.96   (the NARROW band)
     *     kinds 8/9:   0.01 / 0.99
     * Floor (0.0) and envScale (4.0) are identical for all. Band membership gates the process-global dither LCG,
     * so this is invisible dither-free (envelope -> 0 at the edges) yet desyncs the whole frame once dither is on. */
    if (ice_kind == 7) {
        s->DitherBandLoAnchor = float_from_bits(0x3D23D70Au);   /* 0.04 */
        s->DitherBandHiAnchor = float_from_bits(0x3F75C28Fu);   /* 0.96 */
    } else {
        s->DitherBandLoAnchor = float_from_bits(0x3C23D70Au);   /* 0.01 */
        s->DitherBandHiAnchor = float_from_bits(0x3F7D70A4u);   /* 0.99 */
    }
    s->LcgScale           = float_from_bits(0x337FFFFFu);   /* 2^-24 */
    s->LcgBias            = -0.5f;
    s->LcgNegFixup        = 0.0f;
    return s;
}

void ice_state_free(IceState *s) { if (s) { free(s->densityLut); free(s); } }

File: src/IceSetup.cs
using System;

// IceSetup.cs -- setup: build a fully-populated IceState (scalar config, density LUT, and the fixed constants
// the core reads) from the image geometry.
//
// Each value is either [DERIVED] from geometry / a formula, or [FIXED] -- a target-8 constant written as its
// exact 32-bit pattern. The [FIXED] values are scanner-profile coefficients that are invariant for target 8;
// they are embedded as exact bit patterns so the arithmetic downstream is reproducible.
//
// The ring / plane / staging buffers are sized from W and owned separately (see IceBuffers); this file only
// fills the scalar config, the density LUT, and the constants. Config fields are written by NAME (Cfg.*)
// where one exists; the remaining bare-hex offsets are structural (ring seeds, intermediate stores).

public static class IceSetup
{
    // ICE quality: false = Normal (default), true = Fine. Set by openice.cs's -fine. (See docs/parameters.md.)
    public static bool Fine = false;

    // Reconstruction target ("kind"): which scanner reconstruction profile to use. 8 = LS-5000 (default), 7/9 = the
    // two other targets. Only the ReconCoefBase matrix and BandLookaheadRows differ between kinds. Set by
    // openice.cs's -kind. (This C# build is the readable reference; use the C build when byte-exactness matters.)
    public static int Kind = 8;

    // [kind index 0=7,1=8,2=9][channel R,G,B][IR-slope k + 3 bands x {lo,hi}] -- exact 32-bit float patterns.
    static readonly uint[][][] ReconCoefByKind = {
        new[] {  // kind 7
            new uint[]{0x3F8CCCCD,0x3FAE147B,0x3FA8F5C3,0x3FAF5C29,0x3FA66666,0x3FAB851F,0x3FA00000},
            new uint[]{0x3F8CCCCD,0x3FAF5C29,0x3FA66666,0x3FACCCCD,0x3FA51EB8,0x3FA66666,0x3F9EB852},
            new uint[]{0x3F8CCCCD,0x3FAB851F,0x3FA00000,0x3FA8F5C3,0x3FA00000,0x3FA00000,0x3F9AE148} },
        new[] {  // kind 8 (LS-5000, default)
            new uint[]{0x3F8CCCCD,0x3F9AE148,0x3F8B851F,0x3F95C28F,0x3F8A3D71,0x3F851EB8,0x3F75C28F},
            new uint[]{0x3F8CCCCD,0x3F9D70A4,0x3F90A3D7,0x3F91EB85,0x3F866666,0x3F6E147B,0x3F570A3D},
            new uint[]{0x3F8CCCCD,0x3F90A3D7,0x3F851EB8,0x3F8A3D71,0x3F828F5C,0x3F7851EC,0x3F63D70A} },
        new[] {  // kind 9
            new uint[]{0x3F800000,0x400D70A4,0x4005C28F,0x400AE148,0x40051EB8,0x40028F5C,0x3FFAE148},
            new uint[]{0x3F800000,0x400EB852,0x400851EC,0x4008F5C3,0x40033333,0x3FF70A3D,0x3FEB851F},
            new uint[]{0x3F800000,0x400851EC,0x40028F5C,0x40051EB8,0x400147AE,0x3FFC28F6,0x3FF1EB85} }
    };
    static readonly int[] BandLookaheadByKind = { 950, 1600, 2500 };   // kinds 7, 8, 9
    static int KindIndex() { int i = Kind - 7; return (i < 0 || i > 2) ? 1 : i; }

    static void WriteI   (IceState s, int off, int v)      { BitConverter.GetBytes(v).CopyTo(s.config, off); }
    static void WriteBits(IceState s, int off, uint bits)  { BitConverter.GetBytes(bits).CopyTo(s.config, off); }   // exact float bits
    static void WriteF   (IceState s, int off, float v)    { BitConverter.GetBytes(v).CopyTo(s.config, off); }
    static void WriteU16 (IceState s, int off, ushort v)   { BitConverter.GetBytes(v).CopyTo(s.config, off); }
    static float FloatFromBits(uint bits)                  { return BitConverter.ToSingle(BitConverter.GetBytes(bits), 0); }
    static uint  BitsOf(float f)                           { return BitConverter.ToUInt32(BitConverter.GetBytes(f), 0); }

    // density-anchor helpers. `(int)` truncates toward zero. All three clamp the index to maxIdx.
    static int Idx(float anchor, int maxIdx) { int i = (int)((double)anchor * (double)maxIdx); return i > maxIdx ? maxIdx : i; }
    // LUT[min(max, trunc(anchor*max))] - LUT[max]  (density difference; subtract kept in double)
    static float LutSub(IceState s, float anchor, int maxIdx) { return (float)((double)s.densityLut[Idx(anchor, maxIdx)] - (double)s.densityLut[maxIdx]); }
    // LUT[min(max, trunc(anchor*max))]  (the dust-floor density itself, no subtract)
    static float LutAt (IceState s, float anchor, int maxIdx) { return s.densityLut[Idx(anchor, maxIdx)]; }
    // IRref bias: the density-difference branch on the region anchor vs the fixed statics.
    static float BiasParam(IceState s, float fv1, int maxIdx, double a0, double a4)
    {
        const double a9c = 2.0;
        if (fv1 <= a0) return (a4 < fv1) ? LutSub(s, fv1, maxIdx) : 0f;
        if (a9c <= fv1) return (a4 < fv1) ? LutSub(s, fv1, maxIdx) : 0f;
        double g = a9c - (double)fv1;
        return (float)((double)s.densityLut[maxIdx] - (double)s.densityLut[Idx((float)g, maxIdx)]);   // reversed subtract
    }

    public static IceState Create(int W, int H, int dpi)
    {
        var s = new IceState();
        s.config = new byte[0x1200];

        // ---- geometry / config  [DERIVED from cfg + W] ----
        WriteI(s, Cfg.Dpi, dpi);
        WriteI(s, 0x010, 2);
        WriteI(s, Cfg.TriggerEdgeMode, 0);
        WriteI(s, Cfg.ReconIrRefAdjust, 1); WriteI(s, Cfg.ReconBand0, 1); WriteI(s, Cfg.ReconBand1, 1); WriteI(s, Cfg.ReconBand2, 1);   // per-channel gate coeff flags (IR=1)
        // Cfg.TargetId (0x3B4) is a MISNOMER: NOT the reconstruction kind -- the core reads 8 for ALL kinds 7/8/9.
        // It is the give-up sample cap (maxBelow in the trigger). Writing Kind made -kind 7 give up at >7 / -kind 9
        // at >9, flipping the give-up decision on many pixels. Kind-invariant: always 8.
        WriteI(s, Cfg.ChannelCount, 2); WriteI(s, Cfg.BandCount, 0); WriteI(s, Cfg.TargetId, 8);
        WriteI(s, 0xEC8, 4);              // vertical lookahead margin
        WriteI(s, 0xECC, W + 8);
        WriteI(s, 0xED0, W + 4);
        WriteI(s, Cfg.ImageWidth, W);
        WriteI(s, Cfg.RowCounter, 0);     // image-derived accumulator (seed 0)
        WriteI(s, Cfg.ClampToL3Enable, 1); WriteI(s, Cfg.DrainFlag, 0); WriteI(s, Cfg.ClampToL3Flag, 0);

        // row / ring config (target-8 fixed) [FIXED]
        WriteI(s, Cfg.WarmupRows, 7); WriteI(s, 0xE1C, 6); WriteI(s, 0xE28, 2); WriteI(s, 0xE3C, 3); WriteI(s, 0xE40, 1);
        WriteI(s, 0xE2C, 0); WriteI(s, 0xE30, 1); WriteI(s, 0xE34, 2);   // input channel routing R/G/B -> 0/1/2 (ingest reads these)
        // initial ring slots 0xe44..0xec4 are all 0 -- a zeroed cfg already has them.
        WriteU16(s, Cfg.MaxDensityIdx, 65535); WriteU16(s, 0xD72, 0);   // maxIdx
        WriteI(s, Cfg.BandLookaheadRows, BandLookaheadByKind[KindIndex()]); WriteI(s, Cfg.WeightMinRows, 550); WriteI(s, 0x458, 1);   // band/lookahead window (per reconstruction kind)

        // ---- density LUT: 65536 entries  [FORMULA] LUT[i] = 65535 * log2(i+1) / 16 ----
        // (computed first: the per-scan parameter formulas below read it.)
        s.densityLut = new float[65536];
        double k = 65535.0 / (16.0 * Math.Log(2.0));
        for (int i = 0; i <= 65535; i++)
            s.densityLut[i] = (float)(k * Math.Log(i + 1));

        // ---- scalar constants that are true constants (not scan-derived) ----
        WriteBits(s, 0x3BC, 0x00000000);   // 0.0
        WriteBits(s, Cfg.WeightFloor, 0x3CA3D70A);   // 0.02   weight floor
        WriteBits(s, Cfg.IRref, 0x00000000);         // 0.0    IRref seed (supplied by the pump per-image)
        WriteBits(s, Cfg.MaskOverrideR, 0); WriteBits(s, Cfg.MaskOverrideG, 0); WriteBits(s, Cfg.MaskOverrideB, 0);   // mask overrides off
        WriteU16(s, Cfg.TriggerOffsetL, 4); WriteU16(s, Cfg.TriggerOffsetR, 4); WriteU16(s, 0xF74, 4); WriteU16(s, 0xF76, 4);

        // The three per-channel dither amounts, as exact 32-bit patterns (LS-5000 target-8, scan-invariant).
        WriteBits(s, Cfg.DitherAmtR, 0x3C75C28F);   // 0.015   dither R
        WriteBits(s, Cfg.DitherAmtG, 0x3C75C28F);   // 0.015   dither G
        WriteBits(s, Cfg.DitherAmtB, 0x3CCCCCCD);   // 0.025   dither B

        // Exposure-scaled stage gains (from the host exposure setting, external to the core -- kept fixed).
        WriteBits(s, Cfg.StageGain2, 0x3FA00000);   // 1.25   band 2 gain
        WriteBits(s, Cfg.StageGain1, 0x3FA00000);   // 1.25   band 1 gain
        WriteBits(s, Cfg.StageGain0, 0x3FA00000);   // 1.25   band 0 gain (Normal uses 1.25, not 1.0)
        WriteI(s, Cfg.ClampToL3Enable, 1);          // clamp-to-L3 output path enabled

        // ICE Fine (`-fine`). Between Normal and Fine, the only reconstruction fields that change are the three
        // stage gains (1.25 -> 1.0) and the L3 clamp (on -> off). Setting just these reproduces Fine output
        // byte-exactly. (Fine also zeroes an unread word at 0xF50 and a 192-float table at 0x94; neither affects
        // the reconstruction output.)
        if (Fine || Environment.GetEnvironmentVariable("ICE_FINE") != null)
        {
            WriteBits(s, Cfg.StageGain2, 0x3F800000); WriteBits(s, Cfg.StageGain1, 0x3F800000); WriteBits(s, Cfg.StageGain0, 0x3F800000);   // 1.0
            WriteI(s, Cfg.ClampToL3Enable, 0);
        }

        // ---- per-channel 7-float block at ReconCoefBase (R,G,B): reconstruction coefficients selected by the
        // reconstruction target (Kind, default 8 = LS-5000), at config 0xED8..0xF28; IR-slope k + 3 bands
        // x {loCoef,hiCoef}. Kinds 7/9 are the two other targets (see ReconCoefByKind above).
        uint[][] pc = ReconCoefByKind[KindIndex()];
        for (int c = 0; c < 3; c++)
            for (int i = 0; i < 7; i++)
                WriteBits(s, Cfg.ReconCoefBase + c*0x1C + i*4, pc[c][i]);

        // ---- per-scan parameters: the weight-ramp / bias / dust-floor the pipeline runs on. Each is DERIVED
        // from a scanner-profile anchor (a threshold in [0,1]) via the density LUT. The anchors are the target-8
        // scan config; the truncation `(int)` and the slope kept in double across the store follow the same
        // arithmetic modelling as IceFront/IceBack, so f3c rounds to 1.0 exactly.
        int chans   = 2;            // channel count
        int maxIdx  = 65535;
        double statA = 1.0, statB = 0.0;   // fixed statics (1.0 / 0.0)
        float f30 = LutSub(s, 0.85f,  maxIdx);   // slope region anchor
        // The weight-ramp anchor 0xF30 is per-kind: c4701ac0 (-960.418) for kinds 7/8 but
        // c4702180 (-960.523) for kind 9. WeightSlope=1/f30 follows.
        if (Kind == 9) f30 = FloatFromBits(0xC4702180);
        float f34 = LutSub(s, 1.0f,   maxIdx);   // = LUT[max]-LUT[max] = 0
        float f40 = BiasParam(s, 0.98f, maxIdx, statA, statB);   // IRref bias
        float f4c = LutAt (s, 0.065f, maxIdx);   // dust floor
        WriteBits(s, 0xF30, BitsOf(f30));
        WriteBits(s, 0xF34, BitsOf(f34));
        double fv5 = (double)f30 - (double)f34;                 // kept in double across the store
        if (fv5 == statB) { WriteF(s, Cfg.WeightSlope, -(float)maxIdx); WriteF(s, Cfg.WeightRamp, (float)statA); }
        else { double slope = statA / fv5; WriteF(s, Cfg.WeightSlope, (float)slope); WriteF(s, Cfg.WeightRamp, (float)(slope * (double)f30)); }
        WriteF(s, Cfg.WeightBias, f40);
        WriteF(s, Cfg.MaxDensitySum, (float)((float)chans * (float)maxIdx));     // chans * maxIdx = 131070
        WriteF(s, 0xF48, 0f);                                                    // 0.0
        WriteF(s, Cfg.DustFloor, f4c);

        // ---- fixed constants the core reads directly [exact bits] ----
        s.Zero               = 0f;
        s.RConfGain          = 2f;
        s.MaskClampHi        = 1f;
        s.DitherFloor        = 0f;
        s.DitherEnvScale     = 4f;
        // The dither BAND is per-engine: kind 7 runs the odd reconstruction engine with a narrower band
        // (0.04/0.96) vs 0.01/0.99 for kinds 8/9. Invisible dither-free, but band membership gates the frame-
        // global dither LCG, so the wrong band desyncs the whole dithered frame on the LS-9000.
        if (Kind == 7) {
            s.DitherBandLoAnchor = FloatFromBits(0x3D23D70A);   // 0.04
            s.DitherBandHiAnchor = FloatFromBits(0x3F75C28F);   // 0.96
        } else {
            s.DitherBandLoAnchor = FloatFromBits(0x3C23D70A);   // 0.01
            s.DitherBandHiAnchor = FloatFromBits(0x3F7D70A4);   // 0.99
        }
        s.LcgScale           = FloatFromBits(0x337FFFFF);   // 2^-24
        s.LcgBias            = -0.5f;
        s.LcgNegFixup        = 0f;

        return s;
    }
}
